model/PermissionProvider.php

```diff
@@ -63,29 +67,70 @@ public function getPermissions(User $user, array $resourceIds) {
         return $rights;
     }
 
-    protected function getStateObjects($stateIds) {
+    protected function getStateObjects($stateIds)
+    {
         $wfmodel = $this->getServiceManager()->get(WorkflowModel::SERVICE_ID);
         $stateCache = array();
         foreach ($stateIds as $stateId) {
             if (!empty($stateId)) {
-                $stateCache[$stateId] = $wfmodel->getState($stateId);
+                $state = $wfmodel->getState($stateId);
+                $stateWithRoles = $this->addAtomicRoles($state);
+                $stateCache[$stateId] = $stateWithRoles;
             }
         }
         return $stateCache;
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\PermissionInterface::onResourceCreated()
      */
-    public function onResourceCreated(\core_kernel_classes_Resource $resource) {
+    public function onResourceCreated(\core_kernel_classes_Resource $resource)
+    {
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\permission\PermissionInterface::getSupportedRights()
      */
-    public function getSupportedRights() {
+    public function getSupportedRights()
+    {
         return array('WRITE', 'READ');
     }
-}
\ No newline at end of file
+
+    private function addAtomicRoles(StateObject $state): StateObject
+    {
+        $extensionManager = common_ext_ExtensionsManager::singleton();
+        $workFlowService = $this->getServiceManager()->get(ResourceWorkflowService::SERVICE_ID);
+
+        foreach ($workFlowService->getOption(ResourceWorkflowService::OPTION_EXTENSIONS_WITH_ROLES) as $extensionName) {
+            $manifest = $extensionManager->getExtensionById($extensionName)->getManifest();
+            $includedRoles = $manifest->getIncludedRoles();
+            foreach ($includedRoles as $role => $atomicRoles) {
+                if (in_array($role, $state->getReadRoles())) {
+                    $state = new StateObject(
+                        $state->getId(),
+                        $state->getLabel(),
+                        array_merge($state->getReadRoles(), $atomicRoles),
+                        $state->getWriteRoles(),
+                        $state->getTransitions()
+                    );
+                }
+
+                if (in_array($role, $state->getWriteRoles())) {
+                    $state = new StateObject(
+                        $state->getId(),
+                        $state->getLabel(),
+                        $state->getReadRoles(),
+                        array_merge($state->getWriteRoles(), $atomicRoles),
+                        $state->getTransitions()
+                    );
+                }
+            }
+        }
+
+        return $state;
```

Comment:
Creating a new instance is a bit of a heavy operation.
And here we are creating a new instance on every iteration (sometime twice).

Maybe instead of this, we can collect all necessary data in separate variables (e.g. `$readRoles` and `$writeRoles`) and at the end create a single instance of `StateObject`?

Wdyt?

Reply:
hmmm let me think about it :)

Reply:
updated the new refactored solution and pushed! thanks andrei! :)